ouroboros-network-framework/test/Test/Ouroboros/Network/Server2.hs

```diff
@@ -1670,6 +1798,20 @@ prop_multinode_Sim serverAcc (ArbDataFlow dataFlow) script =
         Nothing -> IdleConn
         Just {} -> ActiveConn
 
+
+-- Right fold of the 'Octopus' which splits its results.
+--
+octoSplit :: Octopus a (Either b c) -> (Octopus a b, Octopus a c)
+octoSplit = go [] []
+  where
+    -- this version seems the fastest one (faster than a strict version, DList
+    -- style or a coinductive definition).
+    go :: [b] -> [c] -> Octopus a (Either b c) -> (Octopus a b, Octopus a c)
+    go bs cs (Nil a) = (Octopus.fromList a bs, Octopus.fromList a cs)
+    go bs cs (Cons (Left  b) o) = go (b : bs)      cs  o
+    go bs cs (Cons (Right c) o) = go      bs  (c : cs) o
```

Comment:
I checked that, the problem with this version is that it's recursive call is not in tail call position.

Reply:
What about building an `Octopus` list type tail recursively instead of a `[a]` one?

Reply:
is this possible?

Reply:
I also checked a `DList` like approach where one builds a function that will build the two octopuses, but the current implementation seemed faster.

Reply:
> is this possible?

Yeah, thinking about it it doesn't seem to be doable due to `Nil a` needing the `a`

Reply:
There's a bug though.  We should either reverse the list before passing it to `Octopus.fromList` or instead of `Octopus.fromList` use a `foldl'` (internally `Octopus.fromList` is a `foldr`).

Reply:
Looks good to me!